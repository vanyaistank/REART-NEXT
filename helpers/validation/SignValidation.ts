import { SignFormValues } from '../../containers/Auth';

export const possibleApiErrorMessages = [
	{
		error:
			'The email address and username you have entered is already registered.',
		field: 'email',
	},
	{
		error: 'The email address you have entered is already registered.',
		field: 'email',
	},
	{
		error: 'No such user found.',
		field: 'email',
	},
	{
		error: 'The username you have entered is already registered.',
		field: 'username',
	},
	{
		error: 'The password you have entered is wrong.',
		field: 'password',
	},
];

export const validateSignForm = (values: SignFormValues, isLogin?: boolean) => {
	const errors: any = {};

	const requiredField = (field: string): string => `${field} is required`;
	const minimumLength = (field: string, num?: number): string =>
		`${field} must contain minimum ${num || 8} characters`;
	const maximumLength = (field: string, num?: number): string =>
		`${field} must contain maximum ${num || 24} characters`;

	if (!isLogin) {
		if (!values.username) {
			errors.username = requiredField('Username');
		} else if (!/^[0-9a-zA-Z]+$/.test(values.username)) {
			errors.username = 'Username must contain only letters and numbers';
		} else if (values.username.length < 8) {
			errors.username = minimumLength('Username');
		} else if (values.username.length > 24) {
			errors.username = maximumLength('Username');
		}
	}

	if (!values.email) {
		errors.email = requiredField('E-mail');
	} else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
		errors.email = 'Please provide a valid E-mail';
	}

	if (!values.password) {
		errors.password = requiredField('Password');
	} else if (
		!/^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$/.test(
			values.password
		)
	) {
		errors.password = `${minimumLength(
			'Password'
		)}, at least 1 letter, 1 number and 1 special character`;
	} else if (values.password.length > 24) {
		errors.password = maximumLength('Password');
	}

	return errors;
};
