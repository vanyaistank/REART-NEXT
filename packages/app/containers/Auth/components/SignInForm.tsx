import React from 'react';
import { FormikProps } from 'formik';

import { SignFormValues } from '../index';

import {
	Row,
	Col,
	Heading,
	Icon,
	Button,
	Link,
	Modal,
} from '@Components';
import FormInput from '../../../components/Form/FormInput';

interface Props extends FormikProps<SignFormValues> {
	apiErrors: Partial<SignFormValues>;
	isLogin: boolean;
	showModal: boolean;
	toggleModal: () => void;
	switchScreen: () => void; // switch between sign in and sign up screens
	loading: boolean;
}

const SignInForm = (props: Props) => {
	const {
		toggleModal,
		showModal,
		loading,
		handleSubmit,
		handleReset,
	} = props;

	return (
		<Modal
			handleReset={handleReset}
			toggleModal={toggleModal}
			showModal={showModal}
		>
			<Icon
				onClick={toggleModal}
				icon="Cancel"
				absolute
				top="5"
				right="5"
				style={{
					fill: 'white',
				}}
			/>
			<Row justifyContent="center">
				<Col justifyContent="center" size={12}>
					<form style={{ width: '100%' }} onSubmit={handleSubmit}>
						<Col
							centered
							flexDirection="column"
							size={12}
							marginBottom="30px"
						>
							<Heading mono uppercase size="L">
								sign in
							</Heading>
							<Heading mono uppercase size="L">
								to re-art
							</Heading>
						</Col>
						<FormInput
							field="email"
							label="Email"
							{...props}
						/>
						<FormInput
							field="password"
							label="Password"
							{...props}
						/>
						<Col size={12}>
							<Button
								type="submit"
								loading={loading}
								width="100%"
								purple={!loading}
								onClick={() => {}}
							>
								SIGN IN
							</Button>
						</Col>
						<Col
							size={12}
							marginTop="30px"
						>
							<Heading size="S" lighterGray inline>
								Don't have an account?{'  '}
							</Heading>
							<Link href="/signup" size="S">
								SIGN UP
							</Link>
						</Col>
					</form>
				</Col>
			</Row>
		</Modal>
	);
}

export default SignInForm;
