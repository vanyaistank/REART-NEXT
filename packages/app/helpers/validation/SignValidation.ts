import { SignFormValues } from '../../containers/Auth';

export const possibleApiErrorMessages = [
	{
		error:
			'The email address and username you have entered is already registered.',
		field: 'email',
	},
	{
		error: 'The email address you have entered is already registered.',
		field: 'email',
	},
	{
		error: 'No such user found.',
		field: 'email',
	},
	{
		error: 'The username you have entered is already registered.',
		field: 'username',
	},
	{
		error: 'The password you have entered is wrong.',
		field: 'password',
	},
];

export const validateSignForm = (values: SignFormValues, isRegistration?: boolean) => {
	const errors: any = {};

	const requiredField = (field: string): string => `${field} is required`;
	const minimumLength = (field: string, num?: number): string =>
		`${field} must contain minimum ${num || 8} characters`;
	const maximumLength = (field: string, num?: number): string =>
		`${field} must contain maximum ${num || 24} characters`;

	if (isRegistration) {
		if (!values.username) {
			errors.username = requiredField('Username');
		} else if (!/^[0-9a-zA-Z]+$/.test(values.username)) {
			errors.username = 'Username must contain only letters and numbers';
		} else if (values.username.length < 8) {
			errors.username = minimumLength('Username');
		} else if (values.username.length > 24) {
			errors.username = maximumLength('Username');
		}

		const validateName = (value: string, label: string) => {
			const key = label === 'First Name' ? 'firstName' : 'lastName';

			if (!value) {
				errors[key] = requiredField(label);
			} else if (!/^[a-z ,.'-]+$/i.test(value)) {
				errors[key] = `Please provide a valid ${label.toLowerCase()}`;
			} else if (value.length > 30) {
				errors[key] = maximumLength(label);
			}
		};

		validateName(values.firstName, 'First Name');
		validateName(values.lastName, 'Last Name');
	}

	if (!values.email) {
		errors.email = requiredField('E-mail');
	} else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
		errors.email = 'Please provide a valid E-mail';
	}

	if (!values.password) {
		errors.password = requiredField('Password');
	} else if (
		!/^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$/.test(
			values.password,
		)
	) {
		errors.password = `${minimumLength(
			'Password',
		)}, at least 1 letter, 1 number and 1 special character`;
	} else if (values.password.length > 24) {
		errors.password = maximumLength('Password');
	}

	return errors;
};
