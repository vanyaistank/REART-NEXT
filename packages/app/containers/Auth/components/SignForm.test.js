import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import { Formik } from 'formik';
import { validateSignForm } from '@Helpers/validation/SignValidation';

import SignInForm from './SignInForm';

const noop = () => console.log('noop');

const initialValues = isLogin => {
	const values = {
		email: '',
		password: '',
	};

	if (!isLogin) {
		values.username = '';
	}

	return values;
};

const apiErrors = {
	username: null,
	email: null,
	password: null,
};

it('renders Sign In Form with no values', () => {
	const wrapper = shallow(
		<Formik
			initialValues={initialValues(true)}
			onSubmit={noop}
			validate={values => validateSignForm(values, true)}
			render={({ handleChange, ...rest }) => (
				<SignInForm
					apiErrors={apiErrors}
					handleChange={noop}
					toggleModal={noop}
					switchScreen={noop}
					showModal={noop}
					loading={false}
					{...rest}
				/>
			)}
		/>
	).dive();

	expect(wrapper).toMatchSnapshot();
});

it('renders Sign Up Form with no values', () => {
	const wrapper = shallow(
		<Formik
			initialValues={initialValues(false)}
			onSubmit={noop}
			validate={values => validateSignForm(values, false)}
			render={({ handleChange, ...rest }) => (
				<SignInForm
					apiErrors={apiErrors}
					handleChange={noop}
					toggleModal={noop}
					switchScreen={noop}
					showModal={noop}
					loading={false}
					{...rest}
				/>
			)}
		/>
	).dive();

	expect(wrapper).toMatchSnapshot();
});
