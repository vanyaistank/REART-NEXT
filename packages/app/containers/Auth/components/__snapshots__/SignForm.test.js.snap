// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Sign In Form with no values 1`] = `
<Modal
  handleReset={[Function]}
  showModal={[Function]}
  toggleModal={[Function]}
>
  <Icon
    absolute={true}
    icon="Cancel"
    onClick={[Function]}
    pointer={true}
    right="5"
    small={true}
    top="5"
  />
  <styled.div
    justifyContent="center"
  >
    <styled.div
      size={12}
    >
      <form
        onSubmit={[Function]}
      >
        <styled.div
          centered={true}
          marginBottom="50px"
          size={12}
        >
          <Styled(Heading)
            bold={false}
            marginBottom={null}
            marginTop={null}
            mono={true}
            size="L"
            uppercase={true}
          >
            sign up
          </Styled(Heading)>
          <Styled(Heading)
            bold={false}
            marginBottom={null}
            marginTop={null}
            mono={true}
            size="L"
            uppercase={true}
          >
            to re-art
          </Styled(Heading)>
        </styled.div>
        <ColWithMargin>
          <Styled(Input)
            key="email"
            label="email"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="email"
            type="email"
            value=""
          />
        </ColWithMargin>
        <ColWithMargin>
          <Styled(Input)
            key="username"
            label="username"
            name="username"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="username"
            type="text"
          />
        </ColWithMargin>
        <ColWithMargin>
          <Styled(Input)
            key="password"
            label="password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />
        </ColWithMargin>
        <styled.div
          size={12}
        >
          <Styled(Button)
            dark={true}
            loading={false}
            onClick={[Function]}
            type="submit"
            width="100%"
          >
            SIGN UP
          </Styled(Button)>
        </styled.div>
        <ColWithMargin
          marginTop="30px"
        >
          <Styled(Heading)
            bold={false}
            inline={true}
            lighterGray={true}
            marginBottom={null}
            marginTop={null}
            mono={false}
            size="S"
            uppercase={false}
          >
            Already have an account?
              
          </Styled(Heading)>
          <Styled(LinkComponent)
            onClick={[Function]}
            size="S"
          >
            SIGN IN
          </Styled(LinkComponent)>
        </ColWithMargin>
      </form>
    </styled.div>
  </styled.div>
</Modal>
`;

exports[`renders Sign Up Form with no values 1`] = `
<Modal
  handleReset={[Function]}
  showModal={[Function]}
  toggleModal={[Function]}
>
  <Icon
    absolute={true}
    icon="Cancel"
    onClick={[Function]}
    pointer={true}
    right="5"
    small={true}
    top="5"
  />
  <styled.div
    justifyContent="center"
  >
    <styled.div
      size={12}
    >
      <form
        onSubmit={[Function]}
      >
        <styled.div
          centered={true}
          marginBottom="50px"
          size={12}
        >
          <Styled(Heading)
            bold={false}
            marginBottom={null}
            marginTop={null}
            mono={true}
            size="L"
            uppercase={true}
          >
            sign up
          </Styled(Heading)>
          <Styled(Heading)
            bold={false}
            marginBottom={null}
            marginTop={null}
            mono={true}
            size="L"
            uppercase={true}
          >
            to re-art
          </Styled(Heading)>
        </styled.div>
        <ColWithMargin>
          <Styled(Input)
            key="email"
            label="email"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="email"
            type="email"
            value=""
          />
        </ColWithMargin>
        <ColWithMargin>
          <Styled(Input)
            key="username"
            label="username"
            name="username"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="username"
            type="text"
            value=""
          />
        </ColWithMargin>
        <ColWithMargin>
          <Styled(Input)
            key="password"
            label="password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />
        </ColWithMargin>
        <styled.div
          size={12}
        >
          <Styled(Button)
            dark={true}
            loading={false}
            onClick={[Function]}
            type="submit"
            width="100%"
          >
            SIGN UP
          </Styled(Button)>
        </styled.div>
        <ColWithMargin
          marginTop="30px"
        >
          <Styled(Heading)
            bold={false}
            inline={true}
            lighterGray={true}
            marginBottom={null}
            marginTop={null}
            mono={false}
            size="S"
            uppercase={false}
          >
            Already have an account?
              
          </Styled(Heading)>
          <Styled(LinkComponent)
            onClick={[Function]}
            size="S"
          >
            SIGN IN
          </Styled(LinkComponent)>
        </ColWithMargin>
      </form>
    </styled.div>
  </styled.div>
</Modal>
`;
